doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+# Error Categorization
+
+In this document we start to explain how to categorize errors.
+Many errors are unique and can be difficult to categorize so this
+document will act as a guideline.
+
+## Categorization Mindset
+
+Errors along with their unique code only primarily exist because they can be **reacted
+differently on programmatically**. One could argue that an application could always react differently
+based on the error (eg. give yourself write or read permission) but such detailed granularity
+is rarely needed and would come with a lot of maintenance effort. In the previous versions of
+Elektra we were facing many duplicated errors because too many existed
+and developers simply cannot know each and every code.
+
+To get an idea of programmatic reactions take a method call which returns a `Timeout` error. Naturally
+a senseful reaction would be to retry at a later point in time. So
+the reaction here would be to retry with a time based approach. On the other hand
+it does not make sense to differentiate between `No Write Permission` and `No Read Permission` as
+the application just knows that it simply cannot access the desired resource and tells
+the user to grant it.
+
+Categories are leaf based, so you cannot put an error
```

Comment:
> Why having this restriction? This restriction forces us to introduce otherwise not needed categories such as "General Resource".

To avoid that users put it in very general categories

> And the even bigger problem is that it would disallow to add leaves to an already in-use category.

Why? Extending an existing non-leaf category is easy. Just splitting requires more attention. This makes sense in my opinion because if someone thinks of a special case which deserves special handling it deserves its own category and should be clearly distinguishable. New developers might just put it in a non-leaf category despite it deserves this special handling

Reply:
We could give some categories numbers (like "Resource") and others not (like "Validation). Then we could decide on case-per-case if the category is "too general to be used".

> Why?

Let us assume that we need a further category below "Timeout", e.g. "Resolvable By Automatic-Retry". Then we would only move the errors that belong there to this category. Others we would keep directly in "Timeout" (these errors do not need to be touched). I think you understand that going through errors is not something everyone wants to do.

Reply:
> We could give some categories numbers (like "Resource") and others not (like "Validation). Then we could decide on case-per-case if the category is "too general to be used".

We could do that but then developers will use too broad error categories instead of good specific ones. It is really against the mindset of the decision. The branches are 
only used so that api users can react to whole categories of errors instead of specific errors. It also increases your effort to tell devs in every PR if they might want to use a better specific code.

> Others we would keep directly in "Timeout" (these errors do not need to be touched). I think you understand that going through all errors is not something everyone wants to do.

These errors still do not need to be touched because one could rename "Timeout" to another category and change the error code. It is a matter of seconds to do that. (eg. Timeout before: C03000 ... now General Timout: C03100 & "Resolved By Automatic-Retry" C03200)

Reply:
> We could do that but then developers will use too broad error categories instead of good specific ones.

I fully agree here. You are welcome to prohibit to emit errors in some of the (very general) categories.

> These errors still do not need to be touched because one could rename "Timeout" to another category and change the error code.

No, we cannot do that. Because clients rely that they will get the same error category for the same error. If you design it well, it will be possible to make errors more specific. But changing the category, will always destroy the expectations of clients.

Please always remember that Elektra is a library and the *only* purpose is to make the life easier for clients (applications using Elektra).

Such breaking changes (like you do now with the redesign of the error concept) need to be rare exceptions. The concept now needs to be designed that this should never be needed again.

Reply:
> But changing the category, will always destroy the expectations of clients.

The design itself allows exactly for splitting a category without harming existing clients. Clients itself should check for the "branch" (e.g., .startsWith("C03")) and will always be safe no matter we split the categories they receive the same errors.
As in the example above changing e.g., C03000 to C03100 & C03200 (add C03200, change C03000 to C03100 and rename) will not break anything from clients, instead it gives them another ability to only catch more specific errors if they desire (.startsWith("C031")). The "rename and change category" approach I suggested is just a shortcut.

Reply:
Good to hear that you considered this already. Please also write a tutorial how to react to errors correctly. Btw. the text in the description of kdbSet also needs to be updated.

Reply:
What is the reason for this if it is already part of the guideline documentation?

Reply:
There are two different views: people developing Elektra (who set errors) and users (who receive errors). It seems like that the user view is essential to understand how the system will work (e.g. that extending errors works, all users are only allowed to compare with "startsWith").